src/nvme/nbft.c

```diff
@@ -0,0 +1,947 @@
+// SPDX-License-Identifier: LGPL-2.1-or-later
+/*
+ * This file is part of libnvme.
+ * Copyright (c) 2021-2022, Dell Inc. or its subsidiaries.  All Rights Reserved.
+ *
+ * Authors: Stuart Hayes <Stuart_Hayes@Dell.com>
+ *
+ */
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include <arpa/inet.h>
+
+#include "private.h"
+#include "nbft.h"
+#include "log.h"
+
+/*
+ *  ACPI NBFT table structures (spec v0.65)
+ */
+
+#define NBFT_ACPI_SIG		"NBFT"
+
+enum nbft_raw_descriptor_type {
+	NBFT_HEADER,
+	NBFT_CONTROL,
+	NBFT_HOST,
+	NBFT_HFI,
+	NBFT_SSNS,
+	NBFT_SECURITY,
+	NBFT_DISCOVERY,
+	NBFT_HFI_TRANSPORT,
+	RESERVED_8,
+	NBFT_SSNS_EXTENDED_INFO,
+};
+
+typedef struct __attribute__((__packed__)) heap_obj_s {
```

Comment:
- No typedefs, just use struct heap_obj as name throughout the code. 
- remove `__attribute__((__packed__))`, it doesn't do what you expect it does. When all members of the struct are correctly padded, the attribute will remove all alignment information as all accesses are on byte level. `pahole` is your friend

I only mention it here once. There are more of these.

Reply:
Hmm, something to look at more closely, for now I left the attribute there. The thing is that e.g. the `heap_obj_s` struct is 32+16 bits large and that doesn't align well, when embedded in other structs. As the ACPI NBFT table binary format is fixed I don't see a room for adding extra padding or reorganizing the elements. Or maybe I'm missing something...

Reply:
It might not work for all data types, but at least for the nvme data types it doesn't seem to be necessary, as the members are naturally aligned. At least I haven't found one which it wasn't the case.

Reply:
this has been fixed